File: python/aurigraph_sdk/namespaces/nodes.py
"""Namespace for /api/v11/nodes endpoints -- node management."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from aurigraph_sdk.models import NodeInfo, NodeMetrics

if TYPE_CHECKING:
    from aurigraph_sdk.client import AurigraphClient


class NodesApi:
    """Node management operations."""

    def __init__(self, client: AurigraphClient) -> None:
        self._client = client

    def list(self, page: int = 0, page_size: int = 50) -> dict[str, Any]:
        """List network nodes with pagination.

        Args:
            page: Page number (0-based).
            page_size: Number of nodes per page.

        Returns:
            Paginated node list response.
        """
        return self._client._get(f"/nodes?page={page}&pageSize={page_size}")

    def get(self, node_id: str) -> NodeInfo:
        """Get a specific node by ID.

        Args:
            node_id: The node identifier.
        """
        data = self._client._get(f"/nodes/{node_id}")
        return NodeInfo.model_validate(data)

    def metrics(self) -> NodeMetrics:
        """Get aggregate node metrics (totalNodes, activeNodes, etc.)."""
        data = self._client._get("/nodes/metrics")
        return NodeMetrics.model_validate(data)

    def stats(self) -> dict[str, Any]:
        """Get node statistics."""
        return self._client._get("/nodes/stats")

    def register(self, request: dict[str, Any]) -> NodeInfo:
        """Register a new node on the network.

        Args:
            request: Node registration payload (name, type, host, port, etc.).

        Returns:
            The registered node info.
        """
        data = self._client._post("/nodes", request)
        return NodeInfo.model_validate(data)

File: rust/src/namespaces/nodes.rs
//! Node management namespace for `/api/v11/nodes/*` endpoints.
//!
//! Provides node listing, metrics, stats, and registration.
//!
//! # Example
//!
//! ```rust,no_run
//! # async fn example(client: &aurigraph_sdk::AurigraphClient) -> aurigraph_sdk::Result<()> {
//! let nodes = client.nodes().list().await?;
//! println!("Nodes: {}", nodes);
//!
//! let metrics = client.nodes().metrics().await?;
//! println!("Active nodes: {}", metrics["activeNodes"]);
//! # Ok(())
//! # }
//! ```

use crate::client::AurigraphClient;
use crate::error::Result;

/// Nodes API namespace.
///
/// Obtained via [`AurigraphClient::nodes()`].
#[derive(Debug)]
pub struct NodesApi<'a> {
    client: &'a AurigraphClient,
}

impl<'a> NodesApi<'a> {
    pub(crate) fn new(client: &'a AurigraphClient) -> Self {
        Self { client }
    }

    /// List all registered network nodes.
    ///
    /// Calls `GET /api/v11/nodes`.
    pub async fn list(&self) -> Result<serde_json::Value> {
        self.client.get("/nodes").await
    }

    /// Get a specific node by ID.
    ///
    /// Calls `GET /api/v11/nodes/{node_id}`.
    pub async fn get(&self, node_id: &str) -> Result<serde_json::Value> {
        self.client
            .get(&format!("/nodes/{}", node_id))
            .await
    }

    /// Get aggregate node metrics (total, active, validator count, status).
    ///
    /// Calls `GET /api/v11/nodes/metrics`.
    pub async fn metrics(&self) -> Result<serde_json::Value> {
        self.client.get("/nodes/metrics").await
    }

    /// Get node statistics.
    ///
    /// Calls `GET /api/v11/nodes/stats`.
    pub async fn stats(&self) -> Result<serde_json::Value> {
        self.client.get("/nodes/stats").await
    }

    /// Register a new node on the network.
    ///
    /// Calls `POST /api/v11/nodes/register`.
    pub async fn register(
        &self,
        name: &str,
        node_type: &str,
        host: &str,
        port: u16,
    ) -> Result<serde_json::Value> {
        let body = serde_json::json!({
            "name": name,
            "type": node_type,
            "host": host,
            "port": port,
        });
        self.client.post("/nodes/register", &body).await
    }
}
